### PR TITLE
GOSSIP-62 Implement Crdt PN-Counter

### PR DESCRIPTION
**Pull request by Terry-Weymouth**
I think I got this right, but was a little unsure about the application of GrowOnlyCounter.Builder and storing a pointer to the GossipManager in PNCounter instance. Check me there, and otherwise.

Just reviewed the actual file changes and it looks like I did a control-shift-F in the pom.xml. Reformatted it, sorry about that. I'll fix that and push to this PR.

### CODE REVIEW

**Comment by Terry-Weymouth** [comment]
Tests passing in Travis.

File: gossip-base/src/main/java/org/apache/gossip/crdt/PNCounter.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gossip.crdt;
+
+import org.apache.gossip.manager.GossipManager;
+
+public class PNCounter implements CrdtCounter<Long, PNCounter> {
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/crdt/PNCounter.java]: You are not following the code-style. It is 2 space indent. We allow people top use tabs in the xml files.

File: gossip-base/src/main/java/org/apache/gossip/crdt/PNCounter.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	private GrowOnlyCounter nCount;
+	
+	public PNCounter(GossipManager manager) {
+		myManager = manager;
+		pCount = new GrowOnlyCounter(manager);
+		nCount = new GrowOnlyCounter(manager);
+	}
+
+	@Override
+	public PNCounter merge(PNCounter other) {
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/crdt/PNCounter.java]: We want to return new instances all the time. This object should be immutable.

File: gossip-base/src/main/java/org/apache/gossip/crdt/PNCounter.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+		long nValue = (long)nCount.value();
+		return pValue - nValue;
+	}
+
+	@Override
+	public PNCounter optimize() {
+		return this;
+	}
+
+	public void increment(long l) {
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/crdt/PNCounter.java]: Again these objects should be immutable, this is why the other Classes use builders and constructor injection

File: gossip-base/src/test/java/org/apache/gossip/crdt/PNCounterTest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		counter.decrement(3L);
+		Assert.assertEquals(6, (long) counter.value());
+
+		counter.decrement(12L);
+		Assert.assertEquals(-6, (long) counter.value());
+	}
+
+	@Test
+	public void oddBallLocalOperationTest() {
+		// mock GossipManager
```
Comment (edwardcapriolo) [inline gossip-base/src/test/java/org/apache/gossip/crdt/PNCounterTest.java]: Do not have comments or unneeded spaces.

Comment (edwardcapriolo) [reply gossip-base/src/test/java/org/apache/gossip/crdt/PNCounterTest.java]: Consider adding an end to end test here https://github.com/apache/incubator-gossip/blob/master/gossip-itest/src/test/java/org/apache/gossip/DataTest.java

Comment (edwardcapriolo) [reply gossip-base/src/test/java/org/apache/gossip/crdt/PNCounterTest.java]: you might need changes in CrdtModule.java  your end to en tests will confirm this.

**Review by edwardcapriolo — CHANGES_REQUESTED** [review]

File: gossip-base/src/test/java/org/apache/gossip/crdt/PNCounterTest.java (hunk 4 above)
Comment (Terry-Weymouth) [reply gossip-base/src/test/java/org/apache/gossip/crdt/PNCounterTest.java]: @edwardcapriolo Thanks for points, etc.. Will review, change. Just wanted to acknowledge

File: gossip-base/src/main/java/org/apache/gossip/crdt/PNCounter.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+import org.apache.gossip.manager.GossipManager;
+
+public class PNCounter implements CrdtCounter<Long, PNCounter> {
+
+	private final GossipManager myManager;
+	private GrowOnlyCounter pCount;
+	private GrowOnlyCounter nCount;
+	
+	public PNCounter(GossipManager manager) {
```
Comment (edwardcapriolo) [inline gossip-base/src/main/java/org/apache/gossip/crdt/PNCounter.java]: These objects need to be serializable/deserializable via Jackson. There is way to do construction injection with Jackson. However you are going to need other constructors. See https://github.com/apache/incubator-gossip/blob/master/gossip-base/src/main/java/org/apache/gossip/crdt/GrowOnlyCounter.java

File: gossip-base/src/test/java/org/apache/gossip/crdt/PNCounterTest.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+		Assert.assertEquals(0, (long) counter.value());
+	}
+	
+	@Test
+	public void localOperationTest() {
+		// mock GossipManager
+		GossipManager manager = mock(GossipManager.class);
+		LocalMember mockMember = mock(LocalMember.class);
+		when(mockMember.getId()).thenReturn("x");
+		when(manager.getMyself()).thenReturn(mockMember);
```
Comment (edwardcapriolo) [inline gossip-base/src/test/java/org/apache/gossip/crdt/PNCounterTest.java]: You have these 4 lines at the beginning of every test. If this happens it is probably best to move this code to make this code into a parent class or use a @before annotation.

**Comment by Terry-Weymouth** [comment]
@edwardcapriolo General question.
Given the extent of changes I am assuming that I should close this pull request; then correct problems and submit a new 'smashed' (single issues) pull request. Does that fit the accepted workflow?

**Comment by Terry-Weymouth** [comment]
@edwardcapriolo @makrusak I noticed that some CRDT class names start with Crdt, e.g. CrdtCounter.java, and others do not, e.g. GrowOnlyCounter.java - is there a preference?

**Comment by edwardcapriolo** [comment]
I am not picky about the names. I chose GrowOnlyCounter over GCounter because in that case I thought it was more self documenting. You do not need to close the PR you can do:
Squash
git rebase -i HEAD~2
force push 
git push -f <your repo> <your branch name>

**Comment by Terry-Weymouth** [comment]
Quick update - Although I thought I had written adequate testing, when I wrote out an example using this new counter and started playing around with it, I ran into a number of problems. So, back to the test-and-fix cycle. A few hours this weekend, perhaps, and a evening or two (when I can) and I should be back on track. However, until then...

**Comment by Terry-Weymouth** [comment]
Ok, got my code working all tests and the example as expected. Problems were with example. Now, I'm having some problems merging DataTest.java, but think I can work through it. Take awhile longer, though.

**Comment by Terry-Weymouth** [comment]
I've somehow messed up this PR! The 20 files in the change list (??!!) include all the changed file from my merge of master into my changes. This appears to be a consequence of the fact that I squashed my branch to it's original "departure"point even though I merged in changes from the changed master (to resolve conflicts) mid-way through the changes.

I'm not sure if this can be fixed. But I'll try. Otherwise, we can close this PR and set up a new one.

**Comment by makrusak** [comment]
You can force push here anything.
For example:
1. Pull latest master version of incubator-gossip
2. git checkout -b GOSSIP-62-final
3. git merge --squash (all and only your new commits)
4. There will be conflicts with CrdtModule and CrdtModule for example. Repair them and commit.
5. Then push: git push terry-remote GOSSIP-62 --force

It should lead to one clean commit in this PR without merge conflicts.
Good luck :)

**Comment by Terry-Weymouth** [comment]
Ok, Got it fixed by (simply?) rebasing to master and force push. Now, only 7 files changed - the ones that I actually changed. **Apologies** for all the white-space changes in DataTest.java - a consequence of running the standard apache format against the previous code.

**Comment by makrusak** [comment]
It's nice! One nuisance is that this PR will conflict with my PR GOSSIP-64 at DataTest. One of us (unlucky) whose commit will be merged later will have to repair it :)

**Comment by edwardcapriolo** [comment]
When thing get bad I do this.. git diff master mybranch > /tmp/patch

then :
git checkout master 
patch -p0 < /tmp/patch

Then make a new commit

**Pull request merged**